File: src/main/java/com/es2/vadebicicleta/externo/commons/repository/NoPersistenceRepository.java
package com.es2.vadebicicleta.externo.commons.repository;

import com.es2.vadebicicleta.externo.commons.repository.manager.LongIdManager;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.*;

public abstract class NoPersistenceRepository<T extends Identificable<Long>> {
    private LongIdManager longIdManager;
    private Map<Long, T> elements;

    @Autowired
    protected NoPersistenceRepository(LongIdManager longIdManager) {
        this.longIdManager = longIdManager;
        this.elements = new HashMap<>();
    }

    public T save(T o){
        if(o.getId() == null){
            Long id = longIdManager.generateId(getClass(), LongIdManager.LongGenerationStrategy.IDENTITY);
            o.setId(id);
        }
        else {
            Optional<T> saved = findById(o.getId());

            if(saved.isEmpty()){
                throw new ElementNotFoundException("Not possible to update element with id: " +
                        o.getId() + ". Does not exists element with this id");
            }
        }

        elements.put(o.getId(), o);
        return o;
    }

    public Iterable<T> saveAll(Iterable<T> entities){
        Iterator<T> it = entities.iterator();
        List<T> returnList = new ArrayList<>();
        while(it.hasNext()){
            T saved = save(it.next());
            returnList.add(saved);
        }

        return returnList;
    }

    public Optional<T> findById(Long id){
        T element = elements.get(id);

        if(element == null){
            return Optional.empty();
        }

        return Optional.of(element);
    }

    public boolean existsById(Long id){
        return elements.containsKey(id);
    }

    public Iterable<T> findAll() {
        return new ArrayList<>(elements.values());
    }

    public Iterable<T> findAllById(Iterable<Long> ids){
        Iterator<Long> it = ids.iterator();
        List<T> returnList = new ArrayList<>();
        while(it.hasNext()){
            Long next = it.next();
            T element = elements.get(next);
            if(element != null){
                returnList.add(element);
            }
        }

        return returnList;
    }

    public long count(){
        return elements.size();
    }

    public void deleteById(Long id){
        elements.remove(id);
    }

    public void delete(T o){
        elements.remove(o.getId());
    }

    public void deleteAllById(Iterable<Long> ids) {
        ids.forEach(this::deleteById);
    }

    public void deleteAll(Iterable<T> entities) {
        entities.forEach(this::delete);
    }

    public void deleteAll(){
        elements.clear();
    }

}

File: src/main/kotlin/com/es2/vadebicicleta/externo/commons/repository/NoPersistenceRepository.kt
package com.es2.vadebicicleta.externo.commons.repository

import com.es2.vadebicicleta.externo.commons.repository.manager.LongIdManager
import com.es2.vadebicicleta.externo.commons.repository.manager.LongIdManager.LongGenerationStrategy
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Repository
import java.util.*
import java.util.function.Consumer

@Repository
abstract class NoPersistenceRepository<T : Identificable<Long?>?> @Autowired protected constructor(private val longIdManager: LongIdManager) {
    private val elements: MutableMap<Long, T> = mutableMapOf()

    fun save(o: T?): T? {
        if(o == null){
            return null
        }

        if (o.id == null) {
            val id : Long = longIdManager.generateId(javaClass, LongGenerationStrategy.IDENTITY)
            o.id = id
        } else {
            val saved = findById(o.id!!)
            if (saved.isEmpty) {
                throw ElementNotFoundException(
                    "Not possible to update element with id: " +
                            o.id + ". Does not exists element with this id"
                )
            }
        }
        elements[o.id!!] = o
        return o
    }

    fun saveAll(entities: Iterable<T>): Iterable<T> {
        val it = entities.iterator()
        val returnList: MutableList<T> = ArrayList()
        while (it.hasNext()) {
            val saved = save(it.next()) ?: continue
            returnList.add(saved)
        }
        return returnList
    }

    fun findById(id: Long): Optional<T & Any> {
        val element = elements[id] ?: return Optional.empty()
        return Optional.of(element)
    }

    fun existsById(id: Long): Boolean {
        return elements.containsKey(id)
    }

    fun findAll(): Iterable<T> {
        return ArrayList(elements.values)
    }

    fun findAllById(ids: Iterable<Long>): Iterable<T> {
        val it = ids.iterator()
        val returnList: MutableList<T> = ArrayList()
        while (it.hasNext()) {
            val next = it.next()
            val element = elements[next]
            if (element != null) {
                returnList.add(element)
            }
        }
        return returnList
    }

    fun count(): Long {
        return elements.size.toLong()
    }

    fun deleteById(id: Long) {
        elements.remove(id)
    }

    fun delete(o: T) {
        elements.remove(o?.id)
    }

    fun deleteAllById(ids: Iterable<Long>) {
        ids.forEach(Consumer { id: Long -> deleteById(id) })
    }

    fun deleteAll(entities: Iterable<T>) {
        entities.forEach(Consumer { o: T -> delete(o) })
    }

    fun deleteAll() {
        elements.clear()
    }
}
